enhancements/subscription-content/subscription-injection.md

```diff
@@ -0,0 +1,368 @@
+   the projected resource CSI driver.
+5. Mount the `entitlement` volume into `/run/secrets/etc-pki-entitlement` for each container in the
+   pod, including init and ephemeral containers.
+6. [optional] Add a `Volume` in the pod for the `yumRepository` referenced in the bundle which
+   sources data via the projected resource CSI driver.
+7. [optional] Mount the `yumRepository` volume into `/run/secrets` for each container in the pod,
+   including init and ephemeral containers.
+
+Per upstream docs, the mutating admission webhook will need to register itself with the cluster and
+run as a service in its own namespace [1]. The operator installing the Subscription Injection
+webhook will need to do the following:
+
+1. Create a `Deployment` that runs the webhook in a highly available mode.
+2. Create a `Service` to expose the webhook pods, with generated TLS cert/key pairs.
+3. Create a `ConfigMap` where the service serving CA can be injected.
+4. Create a `MutatingWebhookConfiguration` that fires the webhook for all created pods.
+5. When the service serving CA in item 3 is injected or updated, the operator needs to update the
+   caBundle provided to the webook configuration in 4.
+
+[1]
+https://kubernetes.io/blog/2019/03/21/a-guide-to-kubernetes-admission-controllers/#mutating-webhook-configuration
+
+#### Bundle API
+
+```yaml
+kind: ClusterBundle
+apiVersion: subscription.openshift.io/v1alpha1
+metadata:
+  name: <name>
+  ... # standard metadata, cluster-scoped so no namespace
+spec:
+  entitlement:
+    name: <secret name with entitlement keys, must be in openshift-config namespace>
+  yumRepository:
+    name: <configMap name with yum repo definition, must be in openshift-config namespace>
+  aggregateToClusterRoles: # cluster roles which the ClusterBundle RBAC should aggregate to.
+                           # This can override the operator default
+  - edit
+  - admin
+status:
+  conditions:
+  - type: Invalid
+    status: "False" # standard k8s conditions, should obey the "abnormal-true" convention
+  - ... # additional informational conditions
+  clusterRole: <name>-subscription-bundle # clusterRole is cluster-scoped
+  shares:
+  - name: <name>-subscription-entitlement # shares are cluster-scoped
+  - name: <name>-subscription-yumRepository
+```
+
+#### Operator API
+
+```yaml
+kind: Operator
+apiVersion: subscription.openshift.io/v1alpha1
+metadata:
+  name: cluster # canonical
+  namespace: openshift-subscription-injection
+spec:
+  defaultClusterBundles: # default cluster bundles installed by the operator
+    - name: rhel7
+      namespace: openshift-config
+      entitlement: etc-pki-entitlement
+      yumRepository: redhat-repo-rhel7
+    - name: rhel8
+      namespace: openshift-config
+      entitlement: etc-pki-entitlement
+      yumRepository: redhat-repo-rhel8
+  defaultAggregatedClusterRoles: # default cluster roles that the ClusterBundle RBAC aggregates to
```

Comment:
not clear to me what purpose these two default fields serve

Reply:
`defaultClusterBundles` can be removed. I added that API when we initially believed that we needed to inject separate `redhat.repo` files into rhel7 and rhel8 containers, and therefore the API should expose configuration that lets OLM configure the default bundles installed by the operator. This is no longer necessary - ubi7/8 containers only need an entitlement mounted in to download subscription-based rpms from the CDN.

`defaultAggregatedClusterRoles` is similar - this would let OLM populate the `aggregateToClusterRoles` field on the default bundle.

For the sake of simplicity I think these can be dropped.

Reply:
yeah i'd prefer to drop these.  in any case it sounds like rather than "Default" they are more like "bootstrap" values (seems like you're trying to avoid a limitation in OLM that it can't just create the clusterbundle directly during install?)

Reply:
Correct - these are more bootstrap values. I think we can establish a meets-min by providing a `cluster` example YAML file that can be displayed in the web console.

Reply:
that sounds better

Reply:
@bparees I've gone ahead and simply dropped the Operator API. As presented today there is nothing there is no configuration for the webhook that should be exposed.